### ISSUE TITLE
twisted.test.test_process.ClosingPipes.test_stdout sometimes fails

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4626]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4626|
|Type|defect|
|Created|2010-08-04 23:17:57Z|
|Branch|https://github.com/twisted/twisted/tree/closingpipes-4626|

This can be seen on at least the hardy32-py2.5-glib2 builder:

```
Traceback (most recent call last):
  File "/var/lib/buildbot/twisted/hardy32-py2.5-glib2/Twisted/twisted/test/test_process.py", line 2386, in _endProcess
    'process reason was %r' % reason)
twisted.trial.unittest.FailTest: process reason was <twisted.python.failure.Failure <class 'twisted.internet.error.ProcessTerminated'>>
```

<details><summary>Searchable metadata</summary>

```
trac-id__4626 4626
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_closingpipes_4626 branches/closingpipes-4626
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1280963877000000 1280963877000000
changetime__1282496694000000 1282496694000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [29914]) Branching to 'closingpipes-4626'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

This seems to be a race condition in the unit test.  I "fixed" it with some sleeps (the best way to fix race conditions, as everyone is well aware).

Another option would be to poll() for HUP instead of relying on an `os.write` to fail with SIGPIPE.  However, that wouldn't work on OS X.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/closingpipes-4626&num_builds=10)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml set owner to @exarkun|
|-|-|

Thanks for fixing this.

Could you please add a comment saying roughly what you say in the comment? "Another option would be ...".

Also, in the tests you could use `assertIn` instead of `assertSubstring`. I personally prefer the former, but will happily concede that it's a matter of taste.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [29938]) Mention an idea for an alternate approach not based on sleeping

refs [#4626](https://github.com/twisted/twisted/issues/4626)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [29939]) Use assertIn instead of assertSubstring

refs [#4626](https://github.com/twisted/twisted/issues/4626)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

Good to land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [29940]) Merge closingpipes-4626

Author: exarkun
Reviewer: jml
Fixes: [#4626](https://github.com/twisted/twisted/issues/4626)

Fix a race condition in `test_stdout` and `test_stderr` which would sometimes
cause the child process to exit before the parent process managed to close its
stdin.  This prevent the test from actually exercising the code path it was
trying to exercise.  The new version adds a lot of delays to the child which
should make the race go the right way almost all of the time.

A fix which completely removes the race would be nice, but appears to be much
more complicated.
